libp2p/protocols/kademlia/dhttypes.nim

```diff
@@ -0,0 +1,60 @@
+import chronicles
+import std/[tables, hashes]
+import results
+
+# TODO: make data formater for log-friendly form.
+
+type EntryKey* = object
+  data*: seq[byte]
+
+type EntryVal* = object
+  data*: seq[byte]
+
+type TimeStamp* = object
+  # Currently a string, because for some reason, that's what is chosen at the protobuf level
+  # TODO: convert between RFC3339 strings and use of integers (i.e. the _correct_ way)
+  ts*: string
+
+type EntryCandidate* = object
+  key*: EntryKey
+  val*: EntryVal
+
+type ValidatedEntry* = object
+  key*: EntryKey
+  val*: EntryVal
+
+type RecordVal* = object
+  val*: EntryVal
+  time*: TimeStamp
+
+## Top tip: add chronicles logs to your implementation
+type EntryValidator* = ref object of RootObj
+method validate*(
+    self: EntryValidator, entry: EntryCandidate
+): bool {.base, raises: [], gcsafe.} =
+  doAssert(false, "unimplimented base method")
+
+type EntrySelector* = ref object of RootObj
+method select*(
+    self: EntrySelector, cand: RecordVal, others: seq[RecordVal]
+): RecordVal {.base, raises: [], gcsafe.} =
+  doAssert(false, "EntrySelection base not implemented")
+
+# TODO: make library public, but hidden to users of library
```

Comment:
resolve before RFR

Reply:
looks like it's either 100% public, or in-file only. Will have to make kademlia.nim much bigger, and just eat the risk of future work not inadvertently abusing items that are accesible when they shouldn't be :crossed_fingers: